### PR TITLE
feat: add authorization and error handling for getInvoiceById

### PR DESCRIPTION
**Pull request by edbert2397**
📌 What does this PR do?
Implements authorization in getInvoiceById to ensure users can only access their own invoices.
Adds comprehensive error handling, including validation for invalid invoice IDs.
Improves test coverage by handling positive, negative, and corner cases.

✅ Positive Cases:
✔ Returns 200 OK when a valid invoice exists and belongs to the authenticated user.

❌ Negative Cases - Authorization Failures:
🚫 401 Unauthorized → If req.user is not defined (user not authenticated).
🚫 403 Forbidden → If the invoice does not belong to the authenticated user (partner_id mismatch).

🔴 Negative Cases - Error Handling:
⚠️ 404 Not Found → If the invoice does not exist in the database.
⚠️ 500 Internal Server Error → If an unexpected database error occurs.

⚡ Corner Cases Handled:
🟡 Invalid Input: Returns 400 Bad Request for null, string, or negative ID.

### CODE REVIEW

**Review by tegar-wahyu — COMMENTED** [review]
✅ **LGTM!** PR ini sudah menangani authorization dengan baik dan memiliki error handling yang cukup lengkap. Namun, ada beberapa perbaikan yang bisa meningkatkan keamanan dan kejelasan respons.  

### ⚠ **Saran Perbaikan:**  
**Hindari Logging Detail Sensitif**  
   - Saat terjadi error, `console.error(error)` bisa mencetak stack trace yang sensitif.  
   - **Solusi:** Log hanya `error.message` agar lebih aman.  

   **Ganti ini:**  
   ```js
   console.error("Unhandled error in upload process:", error);
   ```  
   **Menjadi:**  
   ```js
   console.error("Upload process failed:", error.message);
   ```
Sisanya sudah oke👍

File: backend/src/routes/invoiceRoutes.js [hunk 1]
```diff
@@ -8,7 +8,10 @@ router.post(
     authMiddleware,               
     InvoiceController.uploadMiddleware,
     InvoiceController.uploadInvoice
-  );
-  
-router.get('/:id',InvoiceController.getInvoiceById);
```
Comment (kkuratolodon) [inline backend/src/routes/invoiceRoutes.js:13]: Sebelumnya saya sudah coba hit get by id dengan postman, tetapi selalu unauthorized walaupun id dan secret sudah benar: 😭
image
Setelah saya pahami, di route get by id belum menggunakan authMiddleware. ✅
```
router.get('/:id', authMiddleware, InvoiceController.getInvoiceById);
```
Di Postman sudah menghasilkan hasil yang sesuai.
image
Tolong diperbaiki ya, makasih bert! 🚀

**Review by kkuratolodon — CHANGES_REQUESTED** [review]

File: backend/src/controllers/invoiceController.js
Comment (shirinzarqaa) [inline backend/src/controllers/invoiceController.js:1]: Great work on this function! The logic  of _getInvoiceById_ is clear and covers key validation cases, such as checking for a valid `id`, ensuring authentication, and handling different error scenarios properly.  

One improvement that could make this even better is adding docstrings  for better readability and maintainability. This aligns with best practices, such as those discussed in [[this article](https://medium.com/@sydneymarierunkle/the-case-for-documenting-code-a069e15f5cf7)](https://medium.com/@sydneymarierunkle/the-case-for-documenting-code-a069e15f5cf7), which highlights how documentation helps new contributors, improves code reputation, and encourages better design thinking.

**Review by shirinzarqaa — APPROVED** [review]

File: backend/src/routes/invoiceRoutes.js (hunk 1 above)
Comment (edbert2397) [reply backend/src/routes/invoiceRoutes.js:13]: Thanks irfan, it's already resolved.
Screenshot 2025-03-09 140248

**Comment by edbert2397** [comment]
> ✅ **LGTM!** PR ini sudah menangani authorization dengan baik dan memiliki error handling yang cukup lengkap. Namun, ada beberapa perbaikan yang bisa meningkatkan keamanan dan kejelasan respons.
> 
> ### ⚠ **Saran Perbaikan:**
> **Hindari Logging Detail Sensitif**
> 
> * Saat terjadi error, `console.error(error)` bisa mencetak stack trace yang sensitif.
> * **Solusi:** Log hanya `error.message` agar lebih aman.
> 
> **Ganti ini:**
> 
> ```js
> console.error("Unhandled error in upload process:", error);
> ```
> 
> **Menjadi:**
> 
> ```js
> console.error("Upload process failed:", error.message);
> ```
> 
> Sisanya sudah oke👍

**Review by shirinzarqaa — APPROVED** [review]
Code has been refined. Nice work!

**Review by kkuratolodon — APPROVED** [review]
Terima kasih Edbert sudah memperbaiki kodenya. Sekarang sudah berjalan dengan baik!!! 🚀

**Review by suryata — APPROVED** [review]
So far ini merupakan kerja yang sangat baik. Sudah dihubungkan dengan fitur authorization yang sudah saya buat sebelumnya dan berjalan dengan baik di lokal. Good job and  Keep it up **edbert halim**!

**Review by asteriskzie — APPROVED** [review]
Fitur sudah diimplementasikan dengan baik, test case mengcover berbagai skenario. Good job edbert halim 🚀
